drivers/cc1200/cc1200-rxtx.c

```diff
@@ -0,0 +1,430 @@
+    }
+    DEBUG("\n");
+#endif
+
+    int size = pkt->length + 1;
+
+    int left = size - dev->pkt_buf.pos;
+
+    if (!left) {
+        dev->cc1200_statistic.raw_packets_out++;
+
+        LOG_DEBUG("cc1200: packet successfully sent.\n");
+        int tx_bytes = cc1200_get_reg_robust(dev, CC1200_NUM_TXBYTES);
+
+        while (tx_bytes > 0) {
+            DEBUG("%s: %d bytes to send...\n", __func__, tx_bytes);
+            xtimer_usleep(50);
+            tx_bytes = cc1200_get_reg_robust(dev, CC1200_NUM_TXBYTES);
+        }
+
+        //xtimer_spin(xtimer_ticks_from_usec(300));
+        cc1200_switch_to_rx(dev);
+        return;
+    }
+
+    int fifo = CC1200_TX_FIFO_SIZE - cc1200_get_reg_robust(dev, CC1200_NUM_TXBYTES);
+
+    int status = cc1200_strobe(dev, CC1200_SNOP);
+    if ((status & 0x70) == 0x70) {
+        DEBUG("%s:%s:%u fifo over/underflow!?\n", RIOT_FILE_RELATIVE, __func__, __LINE__);
+        _tx_abort(dev);
+        return;
+    }
+
+    int to_send = left > fifo ? fifo : left;
+
+
+    DEBUG("%s: left=%d, to_send=%d, fifo=%d, pos=%d\n",
+          __func__, left, to_send, fifo, dev->pkt_buf.pos);
+
+    /* Flush TX  and calibrate Crystal */
+    cc1200_strobe(dev, CC1200_SFTX);
+    cc1200_strobe(dev, CC1200_SFSTXON);
+
+    /* Write packet into TX FIFO */
+    cc1200_writeburst_reg(dev, CC1200_TXFIFO, ((char *)pkt) + dev->pkt_buf.pos, to_send);
+    dev->pkt_buf.pos += to_send;
+
+    if (left == size) {
+        /* Switch to TX mode */
+        cc1200_strobe(dev, CC1200_STX);
+        //xtimer_usleep(200);
+    }
+
+    if (to_send < left) {
+        /* set GDO2 to 0x2 -> will deassert at TX FIFO below threshold */
+        cc1200_write_reg(dev, CC1200_IOCFG2, CC1200_CFG_TXFIFO_THR);
+        gpio_irq_enable(dev->params.int_pin);
+    }
+    else {
+        /* set GDO2 to 0x6 -> pin 2 will deassert at end of TX packet */
+        cc1200_write_reg(dev, CC1200_IOCFG2, CC1200_CFG_PKT_SYNC_RXTX);
+        gpio_irq_enable(dev->params.int_pin);
+    }
+}
+
+void cc1200_isr_handler(cc1200_t *dev, void (*callback)(void *), void *arg)
+{
+    uint8_t rxbytes;
+
+    switch (dev->radio_state) {
+        case RADIO_RX:
+            rxbytes = cc1200_read_reg(dev, CC1200_NUM_RXBYTES);
+            //DEBUG("radio rx\n");
+            if (gpio_read(dev->params.int_pin) | (rxbytes > 0)) {
+                //DEBUG("cc1200_isr_handler((): starting RX\n");
+                dev->isr_cb = callback;
+                dev->isr_cb_arg = arg;
+                _rx_start(dev);
+
+                /*
+                 * Don't know, why this waiting is necessary, but the driver
+                 * blocks without this. Alternatively, the delay can also be
+                 * located in  _netdev_cc1200_isr() or _isr(), but this here
+                 * is more specific. Also, the delay can be placed before or
+                 * after, but not in _rx_start()...
+                 */
+                xtimer_spin(xtimer_ticks_from_usec(800));
```

Comment:
:/

Reply:
Yes. Unfortunately there seems to be some kind of race condition. I haven't figured out the reason but the spin can be decreased to `50` at least.